docs: create automerge configuration/troubleshooting guide (#7724)
renovatebot/renovate: docs/usage/automerge-configuration.md

      "packagePatterns": [ "eslint" ],
      "groupName": "eslint",
      "schedule": ["before 2am on monday"],
      "automerge": true,
      "automergeType": "branch"
    }
  ]
```

Have you come up with a rule that you think others would benefit from?
How about a PR back to [renovate-config](https://github.com/singapore/renovate-config) with the above rule named `":automergeEslintWeekly"` ?

## Frequent problems and how to resolve them

### Automerge config not correct to begin with

What are the most common mistakes people make when doing the initial configuring?

### Branch vs PR

Those of you familiar with GitHub might note that even if you automerge PRs, you are still going to get notifications (noise) anyway - one when the PR is created and another when it is merged.
For this reason we recommend you consider setting `automergeType=branch` which will mean:

- Renovate first creates a branch and no PR
- If tests pass, Renovate pushes a commit directly to `master` without PR
- If tests fail, Renovate raises a PR for you to review

The result is that passing updates are essentially "silent" - the only sign of them are the commits to your `master` branch.

### Absence of tests

> If you have no tests but still want Renovate to automerge, you need to add "requiredStatusChecks": null to your configuration.

Reviewer: We should note that this is a really bad idea to automerge without Tests.

Author: I agree that it's silly to automerge if you don't have good test coverage.
How about something like this instead?

```suggestion
We recommend that you only turn on automerge once you have good test coverage.
Having good test coverage will prevent you from automerging dependencies that break your build.

If you still want to turn on automerge, despite having no or lackluster tests, you can add "requiredStatusChecks": null to your configuration.
```

Author: Rhys made some changes to the text, so I'm assuming that's the way it should be worded.

### Branch protection rules

What branch protection rules specifically are causing problems?

### Required reviews

> I think we want users to use the Renovate-approve and renovate-approve-2 bots?

Reviewer: Those are an option if reviews are required

Probably should add links to those on the GitHub marketplace?

> Do we have bots for other platforms or only for GitHub?

Reviewer: Only those two for github.

---
title: Automerge configuration/troubleshooting
description: Learn all about Renovate's automerge functionality here
---

# Introduction

> Automerging is a Renovate feature that can save you a lot of time/noise directly, while also benefiting grouping and scheduling.
> In short: it means that Renovate can merge PRs or even branches itself if they pass your tests.

Reviewer 2: ```suggestion
Automerging is a Renovate feature that can let you automate certain dependency updating completely.
```

For this reason we recommend you consider setting `automergeType=branch` which will mean:

- Renovate first creates a branch and no PR
- If tests pass, Renovate pushes a commit directly to `master` without PR
- If tests fail, Renovate raises a PR for you to review

The result is that passing updates are essentially "silent" - the only sign of them are the commits to your `master` branch.

### Absence of tests

If you have no tests but still want Renovate to automerge, you need to add "requiredStatusChecks": null to your configuration.

### Branch protection rules

What branch protection rules specifically are causing problems?

### Required reviews

If you have turned on "Require pull request reviews before merging" in your GitHub branch protection settings, you'll need a way to approve the automerge Renovate pull requests.
The Renovate team has made two apps to automate the approval.

If you require one approval it's enough to install and use the [renovate-approve](https://github.com/apps/renovate-approve) app.

> But, if you require two approvals, then you also need to install and use the [Renovate Approve 2](https://github.com/apps/renovate-approve-2) app.

Author: The `renovate-approve` and `renovate-approve-2` apps do not have a similar user-facing name on the "GitHub app store". It's also a bit weird to link to the appstore, as GitHub themselves seem to only use the Marketplace?
Is this all intended behavior?

Reviewer 2: There's no requirement to add to marketplace. Previously it was only for paid apps anyway

As a general guide, it is recommended that you enable automerge for any types of dependency updates where you would just click Merge anyway.
If there are updates for which you prefer to manually review release notes or code before merging, even when they pass tests, then do not enable automerge for those.

Automerge works particularly well for `devDependencies` as well as for production `dependencies` in projects which have great test coverage.

For example, if you have `jest` or `mocha` as a dependency, and it has an upgrade with passing tests... automerge them!
If you have a linter like `eslint` or `tslint` and its update passes... automerge them!
If you have an API with 100% test coverage and `express` is updated... automerge it!

## Configuration examples

<!-- Internal comment

I think the endgame here should be:

- Each example use case is fully realised, and should be copy/pasteable and work.
- The example should explain the reasoning/thinking behind each setting.
- The example should have links to each setting, so that the docs can be consulted.

-->

### Automerge lock file maintenance

The lowest risk type of update to automerge is probably `lockFileMaintenance`.

> When Renovate performs lock file maintenance, it means that it leaves the project dependency definitions unchanged, but refreshes the lock file completely so that latest versions according to package file constraints are installed.

Author: ```suggestion
When Renovate performs lock file maintenance, it leaves the project dependency definitions unchanged, but refreshes the lock file completely so that the latest versions according to the package file constraints are installed.
```

Here is an example of automerging lock file maintenance:

```json
{
  "lockFileMaintenance": {
    "enabled": true,
    "automerge": true
  }
}
```

### Automerge lint tool updates

Automerging lint tool updates can be a real time-saver.

> Sometimes an update to a lint tool or plugin definition can cause tests to fail, and that is usually deliberate/intentional because the lint authors have added a new validation that you need to adhere to.

Author: ```suggestion
Sometimes an update to a lint tool or plugin definition can cause tests to fail, and that is usually deliberate/intentional because the lint authors have added a new rule that you need to adhere to.
```

Do we like `validation` or `rule` better here?

## Branch vs PR automerging

Even if you automerge PRs, you are likely to still get notification noise - one when the PR is created and another when it is merged.
For this reason we recommend you consider setting `automergeType=branch` which will mean:

- Renovate first creates a branch and no PR
- If tests pass, Renovate pushes a commit directly to the base branch without PR
- If tests fail, Renovate raises a PR for you to review

The result is that passing updates are essentially "silent" - the only sign of them are the commits to your `master` branch.
If you have enabled branch protection which prevents Renovate from automerging directly to the base branch, then this won't work and you should stick with the default PR-based automerging instead.

## Assignees and Reviewers

When automerge is enabled on a PR, Renovate will _not_ add assignees or reviewers at PR creation time, in order to decrease notifications noise a little.
If tests subsequently _fail_, making automerge not possible, then Renovate will then add the configured assignees and/or reviewers.

## Frequent problems and how to resolve them

### Automerge not enabled correctly in config

> Sometimes, the reason Renovate is not automerging a PR is because of a configuration mistake.
> You can confirm that Renovate knows you want it to automerge by seeing if the PR body includes the text "Automerge: Enabled".
> If you see "Automerge: Disabled by config" then it means you need a config change for automerge to work.

Author: ```suggestion
Sometimes, the reason Renovate is not automerging a PR is because of a configuration mistake.
You can confirm that Renovate knows you want it to automerge by checking if the PR body includes the text "Automerge: Enabled".
If you see "Automerge: Disabled by config" then it means you need a config change for automerge to work.
```

### Absence of tests

By default, Renovate will not automerge until it sees passing status checks / check runs for the branch.
If you have no tests but still want Renovate to automerge, you need to add `"requiredStatusChecks": null` to your configuration.

> However, we strongly recommend you should have tests in any project where you are regularly updating dependencies.

Author: ```suggestion
However, we strongly recommend you have tests in any project where you are regularly updating dependencies.
```

Here is an example of automerging lock file maintenance:

```json
{
  "lockFileMaintenance": {
    "enabled": true,
    "automerge": true
  }
}
```

### Automerge lint tool updates

Automerging lint tool updates can be a real time-saver.
Sometimes an update to a lint tool or plugin definition can cause tests to fail, and that is usually deliberate/intentional because the lint authors have added a new validation that you need to adhere to.
However, in many cases the new version(s) will pass tests, and if so then there's really nothing else to consider before merging, so they may as well be automerged:

```json
{
  "packageRules": [
    {
      "depTypeList": ["devDependencies"],
      "packagePatterns": ["lint", "prettier"],
      "automerge": true
    }
  ]
}
```

### Automerge non-major updates

Non-major updates in SemVer ecosystems shouldn't have breaking changes (if they follow the spec), therefore many users enable automerge for these too:

```json
{
  "packageRules": [
    {
      "updateTypes": ["minor", "patch"],
      "matchCurrentVersion": "!/^0/",
      "automerge": true
    }
  ]
}

> ```

Author: @rarkins https://docs.renovatebot.com/configuration-options/#matchcurrentversion says you can use this:

```json
{
  "packageRules": [
    {
      "packagePatterns": ["io.github.resilience4j"],
      "allowedVersions": "!/^0\\./"
    }
  ]
}
```

So I was wondering if the new example in the automerge guide is the right/best way? Or is there more than one valid way to do the exclusion of pre-1.0.0 versions?

What's the real difference between `"matchCurrentVersion": "!/^0/"` and `"allowedVersions": "!/^0\\./"`?
I'm confused because it seems matchCurrentVersion and allowedVersions are doing kinda the same thing?

The docs say that `matchCurrentVersion` deals with semver versions or semver range, and that `allowedVersions` is for capturing allowed versions for dependencies?

Also the block talking about `matchCurrentVersion` suddently use `allowedVersions` in a code example...
Should I open a new issue to discuss this further?

Reviewer 2: These are two different things. The first one is for automerging versions only when the existing version is >=1.0.0. The one you mention is a little weird and says don't allow NEW versions starting with `0`.

`matchCurrentVersion` is a matching rule, used to decide whether this package rule applies. `allowedVersions` is an instruction on how to filter the list of possible versions to update to.

Author: > `matchCurrentVersion` is a matching rule, used to decide whether this package rule applies. `allowedVersions` is an instruction on how to filter the list of possible versions to update to.

Ok. :+1: I'll resolve this conversation.

@rarkins, can you please also take a look at the review comments I've left in this [PR's file tab](https://github.com/renovatebot/renovate/pull/7724/files) and commit what you like best?

---
title: Automerge configuration and troubleshooting
description: Learn all about Renovate's automerge functionality here
---

# Introduction

Automerging is a Renovate feature that can enable you to fully automate upgrading of certain dependencies.
When enabled, it means Renovate will attempt to merge the proposed update once it has passed tests.

As a general guide, it is recommended that you enable automerge for any types of dependency updates where you would just click Merge anyway.
If there are updates for which you prefer to manually review release notes or code before merging, even when they pass tests, then do not enable automerge for those.

> Automerge works particularly well for `devDependencies` as well as for production `dependencies` in projects which have great test coverage.

Author: Hi @rarkins, I noticed you changed the text here, but I lifted this from the current "deep dive" on noise reduction.
You might want to change that docs as well then?

Reviewer 2: One day..
